src/wasm/mod.rs

```diff
@@ -0,0 +1,233 @@
+use alloc::borrow::ToOwned;
+use core::convert::TryFrom;
+use core::fmt;
+
+mod convert;
+mod wasi;
+
+use self::convert::WasmPrimitive;
+
+macro_rules! option_helper {
+    (Some $rt:expr) => {
+        Some($rt)
+    };
+    (Some) => {
+        None
+    };
+}
+
+#[derive(Debug)]
+pub struct Host {
+    pub module: wasmi::ModuleRef,
+    // FIXME: The wasmi crate currently provides no way to determine the
+    // caller's module or memory from a host function. This effectively means
+    // that if a module imports and calls a function from another module, and
+    // that module calls a host function, the memory region of the first module
+    // will be used.
+    //
+    // We will need to either modify the wasmi interpreter and host function API
+    // to pass in function context when calling host methods, or use host
+    // function trampolines which change the `Host` value, to avoid this issue.
+    pub memory: wasmi::MemoryRef,
+}
+
+impl Host {
+    // Create a new host for the given instance.
+    // NOTE: The instance may not have been started yet.
+    fn new(instance: &wasmi::ModuleRef) -> Result<Self, wasmi::Error> {
+        let memory = match instance.export_by_name("memory") {
+            Some(wasmi::ExternVal::Memory(memory)) => memory,
+            _ => {
+                return Err(wasmi::Error::Instantiation(
+                    "required memory export".to_owned(),
+                ))
+            }
+        };
+
+        Ok(Host {
+            module: instance.clone(),
+            memory,
+        })
+    }
+}
+
+macro_rules! host_funcs {
```

Comment:
am i correct that this macro is generating all the wasi host calls (eventually)? it would be nice to have a little more documentation of all the stuff this generates & how it should be invoked (eventually)

Reply:
Yeah. I threw together this macro because I needed to write a boatload of boilerplate to declare even a single host method, and didn't want to have to do it again.

We might actually want to use `witx` files and auto-generate the binding logic (similar to how `wasmtime` does since https://github.com/bytecodealliance/wasmtime/pull/707), but this was much easier to get started with.